### ISSUE TITLE
Broadcast error for complex selections even when arrays the same shape

### ISSUE THREAD

**Issue by sharif1093** [h5py/h5py#1593]
I get `Broadcasting is not supported for complex selections` error
 when I do `arr[:,subarray,...] = block`. Note that both sides are the same shape. Code to reproduce:

```python
try:
    import h5py
    import numpy as np
    f = h5py.File('test.hdf5','w')
    arr = f.create_dataset('data', (5,1000,2), dtype=np.uint8)
    block = np.random.randint(255, size=(5,3,2))
    print(">> Shape of arr selection:", arr[:,[0,2,4],...].shape)
    print(">> Shape of random block:", block.shape)
    arr[:,[0,2,4],...] = block
finally:
    f.close()
```

Output:

```bash
>> Shape of arr selection: (5, 3, 2)
>> Shape of random block: (5, 3, 2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-16-0e6663b2afa7> in <module>
      7     print(">> Shape of arr selection:", arr[:,[0,2,4],...].shape)
      8     print(">> Shape of random block:", block.shape)
----> 9     arr[:,[0,2,4],...] = block
     10 finally:
     11     f.close()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

/usr/local/lib/python3.7/site-packages/h5py/_hl/dataset.py in __setitem__(self, args, val)
    882 
    883         # Perform the write, with broadcasting
--> 884         mspace = h5s.create_simple(selection.expand_shape(mshape))
    885         for fspace in selection.broadcast(mshape):
    886             self.id.write(mspace, fspace, val, mtype, dxpl=self._dxpl)

/usr/local/lib/python3.7/site-packages/h5py/_hl/selections.py in expand_shape(self, source_shape)
    336     def expand_shape(self, source_shape):
    337         if not source_shape == self.array_shape:
--> 338             raise TypeError("Broadcasting is not supported for complex selections")
    339         return source_shape
    340 

TypeError: Broadcasting is not supported for complex selections
```

* Operating System: Ubuntu 16.04.6 LTS in docker
* Python: Python  3.7.7 (system Python on Linux)
* h5py version: 2.10.0
* HDF5 version: 1.12.0
* Numpy version: 1.18.5

**Comment by takluyver** [comment]
Replicated on master - I think I introduced this in #1526. PR coming up.

**Comment by takluyver** [comment]
#1595 should fix it.

**Comment by pkul0005** [comment]
Hi @takluyver & @sharif1093, apologies for my lack of understanding about issues. I am facing the same problem and I'm trying to perform parallel writes to the same file. Is this issue fixed? Why is it still open? Do I need to build h5py again for these fixes to take effect?

**Comment by takluyver** [comment]
If it is this issue you're seeing, it's fixed in that pull request, but that's not merged yet - I'm waiting for one of the other maintainers to review it.

Be aware that parallel writes a more complex topic, though - if you haven't already seen it, look at https://docs.h5py.org/en/latest/mpi.html

**Comment by pkul0005** [comment]
Got it, so once it is merged I can clone the master again and build it. Any idea how long before it is merged?

I have looked at it and quite a few other resources. It has been quite a steep learning curve for me. It would be great to have fancy indexing on multiple axes like ```[[1,2], [1,2]]```. 

Thanks a lot for all your great work!

**Comment by takluyver** [comment]
Yup, that will work when it's merged. We're trying to have a virtual sprint this weekend, so hopefully one of the other maintainers will be able to review it then.

Fancy indexing on multiple axes honestly kind of makes my head hurt, but I have been thinking about adding a way of selecting a list of points, which would be equivalent to fancy indexing on *all* axes. Something like:

```python
dset.points[
    (1, 1),
    (2, 2),
]
```

I'll make a separate issue for this. It might be after 3.0, though.

**Comment by pkul0005** [comment]
That sounds great. Thanks!

**Comment by takluyver** [comment]
I've just opened #1602 about that.
